### ISSUE TITLE
error C2039: 'registry': is not a member of ':spedlog::details'

### ISSUE THREAD

**Issue by feered** [gabime/spdlog#1014]
I recently moved to a new Windows 10 PC and had to reconfigure my environment. Unfortunately I don't have the version of spdlog I was using previously, so I went with the latest available (1.3.1) in vcpkg. I had no issues previously trying to compile.

The application uses a few other dependencies that are all installed via vcpkg and have no issues compiling, so I don't think it is a referencing problem.

I am using Visual Studio 2017 with UCRT version 10.0.17134.0.

These are the errors I am getting:

```
1>e:\github\vcpkg\installed\x64-windows\include\spdlog\spdlog.h(55): error C2039: 'registry': is not a member of 'spdlog::details'
1>e:\github\vcpkg\installed\x64-windows\include\spdlog\details\periodic_worker.h(21): note: see declaration of 'spdlog::details'
1>e:\github\vcpkg\installed\x64-windows\include\spdlog\spdlog.h(55): error C3083: 'registry': the symbol to the left of a '::' must be a type
1>e:\github\vcpkg\installed\x64-windows\include\spdlog\spdlog.h(55): error C2039: 'instance': is not a member of 'spdlog::details'
1>e:\github\vcpkg\installed\x64-windows\include\spdlog\details\periodic_worker.h(21): note: see declaration of 'spdlog::details'
1>e:\github\vcpkg\installed\x64-windows\include\spdlog\spdlog.h(55): error C3861: 'instance': identifier not found
1>e:\github\vcpkg\installed\x64-windows\include\spdlog\spdlog.h(61): error C2039: 'registry': is not a member of 'spdlog::details'
1>e:\github\vcpkg\installed\x64-windows\include\spdlog\details\periodic_worker.h(21): note: see declaration of 'spdlog::details'
1>e:\github\vcpkg\installed\x64-windows\include\spdlog\spdlog.h(61): error C3083: 'registry': the symbol to the left of a '::' must be a type
1>e:\github\vcpkg\installed\x64-windows\include\spdlog\spdlog.h(61): error C2039: 'instance': is not a member of 'spdlog::details'
1>e:\github\vcpkg\installed\x64-windows\include\spdlog\details\periodic_worker.h(21): note: see declaration of 'spdlog::details'
1>e:\github\vcpkg\installed\x64-windows\include\spdlog\spdlog.h(61): error C3861: 'instance': identifier not found
```

**Comment by gabime** [comment]
Okease try use cloned version and see if it vcpkg problem or something else.

**Comment by feered** [comment]
I tried the clone and rebuilt. When I navigate through spdlog's includes, I don't see any issues. Intellisense is showing `registry` as a member.

**Comment by gabime** [comment]
Make sure that the search include paths in the project flags point to the right version.

**Comment by feered** [comment]
Okay, so I managed to find the version of spdlog I was using previously. It looks like 1.0.0 is having no issues compiling with Visual Studio from a fresh install of vcpkg. Only when I go to 1.3.1 does it throw the errors I mentioned above.

**Comment by feered** [comment]
@gabime Okay, going to 1.3.0 indicates that the `spdlog.h` header was missing. I was using the `async.h` from the 1.0.0 release, which required no `spdlog.h` header. I need to go through your documentation to see what has changed since 1.0.0.

However, 1.3.1 no longer errors if `spdlog.h` was not included before the other headers resulting in the very vague build errors in my OP.

Thank you for this wonderful library by the way! It is my go-to logger and I've recommended it to anyone who actively develops in C++.

Cheers.

**Comment by gabime** [comment]
> Thank you for this wonderful library by the way! It is my go-to logger and I've recommended it to anyone who actively develops in C++.

Thats great. Glad you like it and thank you for the feedback.

**Comment by nj2901** [comment]
@feered  @gabime 
I am getting same error when trying the example "Create an asynchronous logger using factory method"  error: no member named 'registry' in namespace 'spdlog::details' . Can you suggest what did you do exactly to resolve this ?

My spdlog version is 1.3.1

**Comment by gabime** [comment]
@nj2901 Make sure your include path point to the right spdlog folder.

**Comment by nj2901** [comment]
it does. i cross - checked it in my netbeans IDE. everything seems to point to correct path. 

Getting blow errors : 
In file included from /usr/local/include/spdlog/async.h:21:
In file included from /usr/local/include/spdlog/details/registry.h:22:
In file included from /usr/local/include/spdlog/sinks/ansicolor_sink.h:9:
/usr/local/include/spdlog/spdlog.h:32:18: error: no member named 'registry' in namespace 'spdlog::details'
        details::registry::instance().initialize_logger(new_logger);
        ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:62:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().initialize_logger(std::move(logger));
    ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:70:21: error: no member named 'registry' in namespace 'spdlog::details'
    return details::registry::instance().get(name);
           ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:76:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().set_formatter(std::move(formatter));
    ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:89:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().set_level(log_level);
    ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:95:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().flush_on(log_level);
    ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:102:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().flush_every(interval);
    ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:108:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().set_error_handler(std::move(handler));
    ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:114:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().register_logger(std::move(logger));
    ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:122:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().apply_all(fun);
    ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:128:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().drop(name);
    ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:134:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().drop_all();
    ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:140:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().shutdown();
    ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:146:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().set_automatic_registration(automatic_registation);
    ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:166:21: error: no member named 'registry' in namespace 'spdlog::details'
    return details::registry::instance().default_logger();
           ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:171:21: error: no member named 'registry' in namespace 'spdlog::details'
    return details::registry::instance().get_default_raw();
           ~~~~~~~~~^
/usr/local/include/spdlog/spdlog.h:176:14: error: no member named 'registry' in namespace 'spdlog::details'
    details::registry::instance().set_default_logger(std::move(default_logger));

**Comment by gabime** [comment]
under ubuntu?

**Comment by nj2901** [comment]
yupp. I am using ubuntu 18.04

**Comment by gabime** [comment]
Ubuntu installs old version. what's the version you see in /user/local/include/spdlog.h?
I suggest remove it and get  fresh from the repo

**Comment by nj2901** [comment]
i got it fresh from the repo . 
In version.h version is shown as 1.3.1 .

**Comment by nj2901** [comment]
any insights ?

**Comment by gabime** [comment]
@feered can you share how you solved this?

**Comment by nj2901** [comment]
for version of spdlog we need to check version.h right ? 

Is there any stable version to which i can rollback and try these asynchronous loggers test programs. 
Also, i guess feered is using Windows as per above conversations.

**Comment by nj2901** [comment]
Version : 

```
nishu@workstation:/usr/local/include/spdlog$ cat version.h 
//
// Copyright(c) 2015 Gabi Melman.
// Distributed under the MIT License (http://opensource.org/licenses/MIT)
//

#pragma once

#define SPDLOG_VER_MAJOR 1
#define SPDLOG_VER_MINOR 3
#define SPDLOG_VER_PATCH 1

#define SPDLOG_VERSION (SPDLOG_VER_MAJOR * 10000 + SPDLOG_VER_MINOR * 100 + SPDLOG_VER_PATCH)
```

**Comment by nj2901** [comment]
@feered It would be great if you can share your insights.

**Comment by gabime** [comment]
1.3.1 is ok

**Comment by feered** [comment]
Have you tried to drop to version 1.2.x?

My issue was related to the include order. `spdlog.h` had to be first.

**Comment by nj2901** [comment]
@feered  i cloned the latest 1.3.1 repo and i am getting above error. 
In which file did you make spdlog.h first ? 
my async.h looks like below : 

```

#include "spdlog/async_logger.h"
#include "spdlog/details/registry.h"
#include "spdlog/details/thread_pool.h"

#include <memory>
#include <mutex>

namespace spdlog {

namespace details {
static const size_t default_async_q_size = 8192;
}

```

**Comment by feered** [comment]
`spdlog.h` is required now. I haven't done my research on the changes in 1.3.1 but I believe logger initialization has changed a bit.

**Comment by binary132** [comment]
I'm also affected by this (macOS 10.14.4.)  This happens when I `#include "spdlog/async.h"` and stops when I stop including it.  The error I get is identical to the one in https://github.com/gabime/spdlog/issues/1014#issuecomment-477511110.

```cpp
#include "spdlog/async.h"
#include "spdlog/sinks/stdout_color_sinks.h"
#include "spdlog/spdlog.h"

// ... main() { ... level = whatever;
    auto  base_logger = spdlog::stderr_color_mt("foo");
    auto& sinks       = base_logger->sinks();
    auto  bar_logger  = std::make_shared<spdlog::logger>("bar", sinks.begin(), sinks.end());
    auto  logger      = std::make_shared<spdlog::logger>("user", sinks.begin(), sinks.end());

    base_logger->set_level(spdlog_levels[level]);
    bar_logger->set_level(spdlog_levels[level]);
    logger->set_level(spdlog_levels[level]);

    logger->info("foo {}", FOO_VERSION);
```

---

```
$ clang --version
Apple LLVM version 10.0.1 (clang-1001.0.46.3)
Target: x86_64-apple-darwin18.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by tt4g** [comment]
@binary132 Does it work in the following `include` order?
If it works, I also think that the include guard is the cause as you thought.

```cpp
#include "spdlog/spdlog.h"
#include "spdlog/sinks/stdout_color_sinks.h"
#include "spdlog/async.h"
```

**Comment by sheridp** [comment]
Just ran into this issue as well.
If spdlog/spdlog.h occurs before stdlog/async.h there is no issue.
However, if I have the following (which is unfortunate, because that's how clang-tidy likes it)
```
#include "spdlog/async.h"
#include "spdlog/sinks/basic_file_sink.h"
#include "spdlog/spdlog.h"
```

> error: 'spdlog::details::registry' has not been declared

Version 1.3.1

**Comment by gabime** [comment]
ok i will repoen this

**Comment by kaiscode** [comment]
This is indeed pretty weird, and can also confirm I have the same issue (specifically with async.h being included before spdlog.h, and also because of clang format which alphabetizes my includes).

I don't know if this is any help, but for me, when I define `SPDLOG_DISABLE_DEFAULT_LOGGER` before the includes, I no longer have these compile errors.

**Comment by kaiscode** [comment]
Alright I'm pretty sure this is why -- In the details/registry.h file, if `SPDLOG_DISABLE_DEFAULT_LOGGER` isn't defined, it will try to include something like wincolor_sink.h, which in turn will include "spdlog.h", which references registry.  Because the preprocessor has already started including registry.h, it won't include it again.  This leads to the contents of spdlog.h referencing registry functions which will only come later in the preprocessed output.

Basically it results in this:
```
async starts including registry.h
registry.h includes wincolor_sink (or ansicolor_sink)
wincolor_sink includes spdlog.h
spdlog.h includes registry.h --> *this is a no-op, which is the problem*
spdlog.h stuff refers to registry class functions -- error!
class registry { blahblahblah } // we wanted this to be before!
```

Does that make sense?

**Comment by gabime** [comment]
Yes. Need to find a way to break this circular include

**Comment by gabime** [comment]
Fixed
